src/memory.cpp

```diff
@@ -35,17 +37,23 @@ SingleArrayMemory::refines(const Memory &other) const {
   auto wRefinement = z3::implies(srcWritable, tgtWritable);
   auto vRefinement = (tgtValue == srcValue);
   auto refinement = z3::implies(tgtSuccess, srcSuccess && wRefinement && vRefinement);
-  return {refinement, {bid, offset}};
+  return {z3::implies(isGlobalBlock(bid), refinement), {bid, offset}};
 }
 
-SingleArrayMemory::SingleArrayMemory(unsigned int numBlocks):
-  Memory(ulog2(numBlocks), numBlocks),
+SingleArrayMemory::SingleArrayMemory(
+    unsigned int numGlobalBlocks, unsigned int nunLocalBlocks):
+  Memory(numGlobalBlocks, nunLocalBlocks, ulog2(numGlobalBlocks + nunLocalBlocks)),
   arrayMaps(ctx.constant("arrayMaps",
     ctx.array_sort(ctx.bv_sort(bidBits), ctx.array_sort(Index::sort(), Float::sort())))),
   writableMaps(ctx.constant("writableMaps",
     ctx.array_sort(ctx.bv_sort(bidBits), ctx.bool_sort()))),
   numelemMaps(ctx.constant("numelemMaps",
-    ctx.array_sort(ctx.bv_sort(bidBits), Index::sort()))) {}
+    ctx.array_sort(ctx.bv_sort(bidBits), Index::sort()))),
+  isGlobalMaps(ctx) {
+  auto bid = ctx.bv_const("bid", bidBits);
+  isGlobalMaps = z3::lambda(bid,
+    z3::ite(z3::ult(bid, numGlobalBlocks), ctx.bool_val(true), ctx.bool_val(false)));
```

Comment:
What about this?
```
isGlobalMaps = [](const smt::expr &bid) {
  return z3::ult(bid, numGlobalBlocks);
};
```

Reply:
This simplifies the encoding by our own.

Reply:
I update it as `isGlobalMaps = z3::lambda(bid, z3::ult(bid, numGlobalBlocks));`